src/rust/engine/src/python.rs

```diff
@@ -131,39 +126,38 @@ impl TypeId {
     py_type.into()
   }
 
-  pub fn as_py_type(&self, py: Python) -> PyType {
+  pub fn as_py_type<'py>(&self, py: Python<'py>) -> &'py PyType {
     // NB: Dereferencing a pointer to a PyTypeObject is safe as long as the module defining the
     // type is not unloaded. That is true today, but would not be if we implemented support for hot
     // reloading of plugins.
-    unsafe { PyType::from_type_ptr(py, self.0 as _) }
+    unsafe { PyType::from_type_ptr(py, self.0) }
   }
 
   pub fn is_union(&self) -> bool {
     let gil = Python::acquire_gil();
     let py = gil.python();
-    let unions = py.import("pants.engine.unions").unwrap();
-    unions
-      .call(py, "is_union", (self.as_py_type(py),), None)
+    let unions_module = py.import("pants.engine.unions").unwrap();
+    let is_union_func = unions_module.getattr("is_union").unwrap();
+    is_union_func
+      .call1((self.as_py_type(py),))
       .unwrap()
-      .extract(py)
+      .extract()
       .unwrap()
   }
 }
 
 impl From<&PyType> for TypeId {
   fn from(py_type: &PyType) -> Self {
-    TypeId(py_type.as_type_ptr() as *mut std::ffi::c_void)
+    TypeId(py_type.as_type_ptr())
   }
 }
 
 impl fmt::Debug for TypeId {
   fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-    let name = {
-      let gil = Python::acquire_gil();
-      let py = gil.python();
-      self.as_py_type(py).name(py).into_owned()
-    };
-    write!(f, "{}", name)
+    Python::with_gil(|py| {
+      let name = self.as_py_type(py).name().unwrap();
+      write!(f, "{}", name)
```

Comment:
I originally had the `write!()` outside the GIL closure but that caused a lifetime issue because `name` is `&str`. I figured it was better to hold the GIL than to clone into a `String`, but not sure? DIdn't microbenchmark